File: app/src/main/java/com/example/rafaelanastacioalves/moby/domain/interactors/Interactor.java
package com.example.rafaelanastacioalves.moby.domain.interactors;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;

public interface Interactor <R extends Interactor.RequestValues> {

    <T extends LiveData> T execute(R requestValue);

    public interface RequestValues {
    }
}

File: app/src/main/java/com/example/rafaelanastacioalves/moby/domain/interactors/Interactor.kt
package com.example.rafaelanastacioalves.moby.domain.interactors

import domain.domain.uiDispatcher
import kotlinx.coroutines.*

abstract class Interactor<out T ,in R : Interactor.RequestValues> {

    abstract suspend fun run(requestValue: R?): T


    suspend operator fun invoke(
            params: R?,
            onResult: (T) -> Unit = {}
    ){
        coroutineScope {
            val backGroundJob = async { run(params) }
            launch (uiDispatcher) {
                onResult(backGroundJob.await())
            }

        }
    }

    interface RequestValues
}
